```
   Compiling style v0.0.1 (file:///C:/buildbot/slave/windows-msvc-dev/build/components/style)
error[E0432]: unresolved import `values::specified::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16359:17
      |
16359 |         pub use values::specified::ScrollSnapPoint as SpecifiedValue;
      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::specified`

error[E0432]: unresolved import `values::computed::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16361:21
      |
16361 |             pub use values::computed::ScrollSnapPoint as T;
      |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::computed`

error[E0432]: unresolved import `values::specified::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16521:17
      |
16521 |         pub use values::specified::ScrollSnapPoint as SpecifiedValue;
      |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::specified`

error[E0432]: unresolved import `values::computed::ScrollSnapPoint`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16523:21
      |
16523 |             pub use values::computed::ScrollSnapPoint as T;
      |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `ScrollSnapPoint` in `values::computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `computed`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16363:69
      |
16363 |         #[inline] pub fn get_initial_value() -> computed_value::T { computed::ScrollSnapPoint::none() }
      |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `specified`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16369:13
      |
16369 |             specified::ScrollSnapPoint::parse(context, input)
      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `specified`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `computed`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16525:69
      |
16525 |         #[inline] pub fn get_initial_value() -> computed_value::T { computed::ScrollSnapPoint::none() }
      |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `computed`

error[E0433]: failed to resolve. Could not find `ScrollSnapPoint` in `specified`
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16531:13
      |
16531 |             specified::ScrollSnapPoint::parse(context, input)
      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `ScrollSnapPoint` in `specified`

error: type `properties::DeclaredValue<'_, _>` cannot be dereferenced
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16407:31
      |
16407 |                         match *value {
      |                               ^^^^^^

error: type `properties::DeclaredValue<'_, _>` cannot be dereferenced
     --> c:\buildbot\slave\windows-msvc-dev\build\target\debug\build\style-56fbb21baf6b3c7d\out/properties.rs:16569:31
      |
16569 |                         match *value {
      |                               ^^^^^^

error: aborting due to previous error(s)

error: Could not compile `style`.
```

--- a/components/style/properties/gecko.mako.rs
+++ b/components/style/properties/gecko.mako.rs
@@ -2232,23 +2232,16 @@ fn static_assert() {
     }
     % endfor
 
-    pub fn set_scroll_snap_points_x(&mut self, v: longhands::scroll_snap_points_x::computed_value::T) {
-        match v.0 {
-            None => self.gecko.mScrollSnapPointsX.set_value(CoordDataValue::None),
-            Some(l) => l.to_gecko_style_coord(&mut self.gecko.mScrollSnapPointsX),
-        };
-    }
+    % for axis in ["x", "y"]:
+        pub fn set_scroll_snap_points_${axis}(&mut self, v: longhands::scroll_snap_points_${axis}::computed_value::T) {
+            match v.repeated() {
+                None => self.gecko.mScrollSnapPoints${axis.upper()}.set_value(CoordDataValue::None),
+                Some(l) => l.to_gecko_style_coord(&mut self.gecko.mScrollSnapPoints${axis.upper()}),
+            };
+        }
 
-    ${impl_coord_copy('scroll_snap_points_x', 'mScrollSnapPointsX')}
-
-    pub fn set_scroll_snap_points_y(&mut self, v: longhands::scroll_snap_points_y::computed_value::T) {
-        match v.0 {
-            None => self.gecko.mScrollSnapPointsY.set_value(CoordDataValue::None),
-            Some(l) => l.to_gecko_style_coord(&mut self.gecko.mScrollSnapPointsY),
-        };
-    }
-
-    ${impl_coord_copy('scroll_snap_points_y', 'mScrollSnapPointsY')}
+        ${impl_coord_copy('scroll_snap_points_' + axis, 'mScrollSnapPoints' + axis.upper())}
+    % endfor
 
     pub fn set_scroll_snap_coordinate<I>(&mut self, v: I)
         where I: IntoIterator<Item = longhands::scroll_snap_coordinate::computed_value::single_value::T>,

--- a/components/style/properties/longhand/box.mako.rs
+++ b/components/style/properties/longhand/box.mako.rs
@@ -650,99 +650,17 @@ ${helpers.predefined_type("animation-delay",
                           spec="https://drafts.csswg.org/css-animations/#propdef-animation-delay",
                           allowed_in_keyframe_block=False)}
 
-<%helpers:longhand products="gecko" name="scroll-snap-points-y" animation_value_type="none"
-                   spec="Nonstandard (https://www.w3.org/TR/2015/WD-css-snappoints-1-20150326/#scroll-snap-points)">
-    use std::fmt;
-    use style_traits::ToCss;
-    use values::specified::LengthOrPercentage;
-
-    pub mod computed_value {
-        use values::computed::LengthOrPercentage;
-
-        #[derive(Debug, Clone, PartialEq)]
-        #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-        pub struct T(pub Option<LengthOrPercentage>);
-    }
-
-    #[derive(Clone, Debug, HasViewportPercentage, PartialEq)]
-    #[cfg_attr(feature = "servo", derive(HeapSizeOf))]
-    pub enum SpecifiedValue {
-        None,
-        Repeat(LengthOrPercentage),
-    }
-
-    impl ToCss for computed_value::T {
-        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
-            match self.0 {
-                None => dest.write_str("none"),
-                Some(ref l) => {
-                    try!(dest.write_str("repeat("));
-                    try!(l.to_css(dest));
-                    dest.write_str(")")
-                },
-            }
-        }
-    }
-    impl ToCss for SpecifiedValue {
-        fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
-            match *self {
-                SpecifiedValue::None => dest.write_str("none"),
-                SpecifiedValue::Repeat(ref l) => {
-                    try!(dest.write_str("repeat("));
-                    try!(l.to_css(dest));
-                    dest.write_str(")")
-                },
-            }
-        }
-    }
-
-    #[inline]
-    pub fn get_initial_value() -> computed_value::T {
-        computed_value::T(None)
-    }
-
-    impl ToComputedValue for SpecifiedValue {
-        type ComputedValue = computed_value::T;
-
-        #[inline]
-        fn to_computed_value(&self, context: &Context) -> computed_value::T {
-            match *self {
-                SpecifiedValue::None => computed_value::T(None),
-                SpecifiedValue::Repeat(ref l) =>
-                    computed_value::T(Some(l.to_computed_value(context))),
-            }
-        }
-        #[inline]
-        fn from_computed_value(computed: &computed_value::T) -> Self {
-            match *computed {
-                computed_value::T(None) => SpecifiedValue::None,
-                computed_value::T(Some(l)) =>
-                    SpecifiedValue::Repeat(ToComputedValue::from_computed_value(&l))
-            }
-        }
-    }
-
-    pub fn parse(context: &ParserContext, input: &mut Parser) -> Result<SpecifiedValue, ()> {
-        if input.try(|input| input.expect_ident_matching("none")).is_ok() {
-            Ok(SpecifiedValue::None)
-        } else if input.try(|input| input.expect_function_matching("repeat")).is_ok() {
-            input.parse_nested_block(|input| {
-                LengthOrPercentage::parse_non_negative(context, input).map(SpecifiedValue::Repeat)
-            })
-        } else {
-            Err(())
-        }
-    }
-</%helpers:longhand>
-
-<%helpers:longhand products="gecko" name="scroll-snap-points-x" animation_value_type="none"
-                   spec="Nonstandard (https://www.w3.org/TR/2015/WD-css-snappoints-1-20150326/#scroll-snap-points)">
-    pub use super::scroll_snap_points_y::SpecifiedValue;
-    pub use super::scroll_snap_points_y::computed_value;
-    pub use super::scroll_snap_points_y::get_initial_value;
-    pub use super::scroll_snap_points_y::parse;
-</%helpers:longhand>
-
+% for axis in ["x", "y"]:
+    ${helpers.predefined_type(
+        "scroll-snap-points-" + axis,
+        "ScrollSnapPoint",
+        "computed::ScrollSnapPoint::none()",
+        animation_value_type="none",
+        products="gecko",
+        disable_when_testing=True,
+        spec="Nonstandard (https://www.w3.org/TR/2015/WD-css-snappoints-1-20150326/#scroll-snap-points)",
+    )}
+% endfor
 
 ${helpers.predefined_type("scroll-snap-destination",
                           "Position",

--- a/components/style/values/computed/gecko.rs
+++ b/components/style/values/computed/gecko.rs
@@ -1,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Computed types for legacy Gecko-only properties.
+
+use values::computed::length::LengthOrPercentage;
+use values::generics::gecko::ScrollSnapPoint as GenericScrollSnapPoint;
+
+/// A computed type for scroll snap points.
+pub type ScrollSnapPoint = GenericScrollSnapPoint<LengthOrPercentage>;

--- a/components/style/values/computed/mod.rs
+++ b/components/style/values/computed/mod.rs
@@ -28,6 +28,8 @@ pub use self::background::BackgroundSize;
 pub use self::border::{BorderImageSlice, BorderImageWidth, BorderImageSideWidth};
 pub use self::border::{BorderRadius, BorderCornerRadius};
 pub use self::image::{Gradient, GradientItem, ImageLayer, LineDirection, Image, ImageRect};
+#[cfg(feature = "gecko")]
+pub use self::gecko::ScrollSnapPoint;
 pub use self::rect::LengthOrNumberRect;
 pub use super::{Auto, Either, None_};
 #[cfg(feature = "gecko")]
@@ -46,6 +48,8 @@ pub mod background;
 pub mod basic_shape;
 pub mod border;
 pub mod image;
+#[cfg(feature = "gecko")]
+pub mod gecko;
 pub mod length;
 pub mod position;
 pub mod rect;

--- a/components/style/values/generics/gecko.rs
+++ b/components/style/values/generics/gecko.rs
@@ -1,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Generic types for legacy Gecko-only properties that should probably be
+//! unshipped at some point in the future.
+
+use std::fmt;
+use style_traits::ToCss;
+
+/// A generic value for scroll snap points.
+#[derive(Clone, Copy, Debug, HasViewportPercentage, PartialEq, ToComputedValue)]
+pub enum ScrollSnapPoint<LengthOrPercentage> {
+    /// `none`
+    None,
+    /// `repeat(<length-or-percentage>)`
+    Repeat(LengthOrPercentage)
+}
+
+impl<L> ScrollSnapPoint<L> {
+    /// Returns `none`.
+    #[inline]
+    pub fn none() -> Self {
+        ScrollSnapPoint::None
+    }
+
+    /// Returns the repeat argument, if any.
+    #[inline]
+    pub fn repeated(&self) -> Option<&L> {
+        match *self {
+            ScrollSnapPoint::None => None,
+            ScrollSnapPoint::Repeat(ref length) => Some(length),
+        }
+    }
+}
+
+impl<L> ToCss for ScrollSnapPoint<L>
+where
+    L: ToCss,
+{
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result
+    where
+        W: fmt::Write,
+    {
+        match *self {
+            ScrollSnapPoint::None => dest.write_str("none"),
+            ScrollSnapPoint::Repeat(ref length) => {
+                dest.write_str("repeat(")?;
+                length.to_css(dest)?;
+                dest.write_str(")")
+            },
+        }
+    }
+}

--- a/components/style/values/generics/mod.rs
+++ b/components/style/values/generics/mod.rs
@@ -16,6 +16,8 @@ use values::specified::url::SpecifiedUrl;
 pub mod background;
 pub mod basic_shape;
 pub mod border;
+#[cfg(feature = "gecko")]
+pub mod gecko;
 pub mod grid;
 pub mod image;
 pub mod position;

--- a/components/style/values/specified/gecko.rs
+++ b/components/style/values/specified/gecko.rs
@@ -1,0 +1,26 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Specified types for legacy Gecko-only properties.
+
+use cssparser::Parser;
+use parser::{Parse, ParserContext};
+use values::generics::gecko::ScrollSnapPoint as GenericScrollSnapPoint;
+use values::specified::length::LengthOrPercentage;
+
+/// A specified type for scroll snap points.
+pub type ScrollSnapPoint = GenericScrollSnapPoint<LengthOrPercentage>;
+
+impl Parse for ScrollSnapPoint {
+    fn parse(context: &ParserContext, input: &mut Parser) -> Result<Self, ()> {
+        if input.try(|i| i.expect_ident_matching("none")).is_ok() {
+            return Ok(GenericScrollSnapPoint::None);
+        }
+        input.expect_function_matching("repeat")?;
+        let length = input.parse_nested_block(|i| {
+            LengthOrPercentage::parse_non_negative(context, i)
+        })?;
+        Ok(GenericScrollSnapPoint::Repeat(length))
+    }
+}

--- a/components/style/values/specified/mod.rs
+++ b/components/style/values/specified/mod.rs
@@ -35,7 +35,8 @@ pub use self::border::{BorderCornerRadius, BorderImageSlice, BorderImageWidth};
 pub use self::border::{BorderImageSideWidth, BorderRadius, BorderSideWidth};
 pub use self::color::Color;
 pub use self::rect::LengthOrNumberRect;
-pub use super::generics::grid::GridLine;
+#[cfg(feature = "gecko")]
+pub use self::gecko::ScrollSnapPoint;
 pub use self::image::{ColorStop, EndingShape as GradientEndingShape, Gradient};
 pub use self::image::{GradientItem, GradientKind, Image, ImageRect, ImageLayer};
 pub use self::length::AbsoluteLength;
@@ -46,6 +47,7 @@ pub use self::length::{MaxLength, MozLength};
 pub use self::position::{Position, PositionComponent};
 pub use self::text::{LetterSpacing, LineHeight, WordSpacing};
 pub use self::transform::{TimingFunction, TransformOrigin};
+pub use super::generics::grid::GridLine;
 
 #[cfg(feature = "gecko")]
 pub mod align;
@@ -54,6 +56,8 @@ pub mod basic_shape;
 pub mod border;
 pub mod calc;
 pub mod color;
+#[cfg(feature = "gecko")]
+pub mod gecko;
 pub mod grid;
 pub mod image;
 pub mod length;

